crates/wasi-common/src/ctx.rs

```diff
@@ -231,27 +234,27 @@ impl WasiCtxBuilder {
 
     /// Provide a File to use as stdin
     pub fn stdin(&mut self, file: File) -> &mut Self {
-        self.stdin = Some(PendingEntry::File(file));
+        self.stdin = Some(PendingEntry::OsHandle(file));
         self
     }
 
     /// Provide a File to use as stdout
     pub fn stdout(&mut self, file: File) -> &mut Self {
-        self.stdout = Some(PendingEntry::File(file));
+        self.stdout = Some(PendingEntry::OsHandle(file));
         self
     }
 
     /// Provide a File to use as stderr
     pub fn stderr(&mut self, file: File) -> &mut Self {
-        self.stderr = Some(PendingEntry::File(file));
+        self.stderr = Some(PendingEntry::OsHandle(file));
         self
     }
 
     /// Add a preopened directory.
     pub fn preopened_dir<P: AsRef<Path>>(&mut self, dir: File, guest_path: P) -> &mut Self {
         self.preopens.as_mut().unwrap().push((
             guest_path.as_ref().to_owned(),
-            Box::new(OsHandle::from(dir)),
+            Box::new(OsDir::try_from(dir).expect("valid OsDir handle")),
```

Comment:
Ah, right! I was meant to do this and somehow forgot. Thanks for pointing this one out!

Reply:
Fixed in 36f07cf. So in order not to force public API changes, instead of throwing an error directly at the preopen callsite (either `preopened_dir` or `preopened_virt`), the creation of the actual instance is now postponed until we build the context object, much like creation of the stdio handles.